File: src/test/scala/io/github/jbellis/brokk/analyzer/CppAnalyzerTest.scala
package io.github.jbellis.brokk.analyzer

import org.junit.jupiter.api.Assertions.*
import org.junit.jupiter.api.Assumptions.assumeTrue
import org.junit.jupiter.api.{BeforeAll, Test}

import java.nio.file.{Files, Path}
import scala.jdk.CollectionConverters.*
import scala.util.Try
// For @JvmStatic, if required by JUnit setup for companion objects in Scala 3.
// However, JUnit 5 typically handles companion object @BeforeAll methods correctly without it.
// If the error persists, one might need to add a dependency providing it or check JUnit/Scala version compatibility.
// import kotlin.jvm.JvmStatic // Example if it were a Kotlin/Java interop scenario.

// Companion object for @BeforeAll, must be top-level or static in Java
object CppAnalyzerTest {
  private val testProjectPath = Path.of("src/test/resources/testcode-cpp").toAbsolutePath()
  private var analyzer: CppAnalyzer = scala.compiletime.uninitialized

  @BeforeAll
  // @JvmStatic // Usually not needed in Scala 3 with JUnit 5 for companion objects.
  // If build still complains about JvmStatic, ensure JUnit Jupiter is correctly configured for Scala.
  def setup(): Unit = {
    val tempFile = Files.createTempFile("brokk-cpp-cpg-", ".bin")
    tempFile.toFile.deleteOnExit()
    analyzer = new CppAnalyzer(testProjectPath, java.util.Collections.emptySet[String](), tempFile)
  }
}

class CppAnalyzerTest {
  // Use the analyzer from the companion object
  private def an: CppAnalyzer = CppAnalyzerTest.analyzer

  private val testProjectPath = Path.of("src/test/resources/testcode-cpp").toAbsolutePath() // Keep for ProjectFile creation

  @Test
  def parseFqNameTest(): Unit = {
    // Helper to call protected method
    def parse(fqn: String, t: CodeUnitType) = an.parseFqName(fqn, t)

    // Global function (c2cpg might prefix with filename or have other conventions)
    // Assuming simple global name if not file-prefixed by CPG
    // Test data has functions like geometry.cpp.global_func
    // and geometry.h contains declarations that might lead to geometry.h.func if not careful with CPG data
    assertEquals(new CodeUnit.Tuple3("geometry_h", "", "global_func_decl_in_header"), parse("geometry_h.global_func_decl_in_header", CodeUnitType.FUNCTION))
    assertEquals(new CodeUnit.Tuple3("geometry_cpp", "", "global_func"), parse("geometry_cpp.global_func", CodeUnitType.FUNCTION))


    // Class / Struct
    assertEquals(new CodeUnit.Tuple3("shapes", "Circle", ""), parse("shapes.Circle", CodeUnitType.CLASS))
    assertEquals(new CodeUnit.Tuple3("", "Point", ""), parse("Point", CodeUnitType.CLASS))

    // Method
    assertEquals(new CodeUnit.Tuple3("shapes", "Circle", "getArea"), parse("shapes.Circle.getArea", CodeUnitType.FUNCTION))
    assertEquals(new CodeUnit.Tuple3("", "Point", "print"), parse("Point.print", CodeUnitType.FUNCTION))

    // Constructor (c2cpg uses class name for constructor)
    assertEquals(new CodeUnit.Tuple3("shapes", "Circle", "Circle"), parse("shapes.Circle.Circle", CodeUnitType.FUNCTION))

    // Field / Member variable
    assertEquals(new CodeUnit.Tuple3("shapes", "Circle", "radius"), parse("shapes.Circle.radius", CodeUnitType.FIELD))
    assertEquals(new CodeUnit.Tuple3("", "Point", "x"), parse("Point.x", CodeUnitType.FIELD))

    // Namespace-only function (not part of a class)
    assertEquals(new CodeUnit.Tuple3("shapes", "", "another_in_shapes"), parse("shapes.another_in_shapes", CodeUnitType.FUNCTION))

    // Fallback cases (where FQN components might not be in CPG as expected)
    assertEquals(new CodeUnit.Tuple3("pkg", "Cls", "method"), parse("pkg.Cls.method", CodeUnitType.FUNCTION), "Fallback: pkg.Cls.method")
    assertEquals(new CodeUnit.Tuple3("pkg", "Cls", ""), parse("pkg.Cls", CodeUnitType.CLASS), "Fallback: pkg.Cls")
    assertEquals(new CodeUnit.Tuple3("pkg", "", "func"), parse("pkg.func", CodeUnitType.FUNCTION), "Fallback: pkg.func")
    // This case needs `Cls` to be a TypeDecl. If not, it would be ("Cls", "", "method") by current fallback.
    // The test implies Cls is treated as a class name even if not in CPG.
    assertEquals(new CodeUnit.Tuple3("", "Cls", "method"), parse("Cls.method", CodeUnitType.FUNCTION), "Fallback: Cls.method")
    assertEquals(new CodeUnit.Tuple3("", "", "func"), parse("func", CodeUnitType.FUNCTION), "Fallback: func (global)")
    assertEquals(new CodeUnit.Tuple3("", "Cls", ""), parse("Cls", CodeUnitType.CLASS), "Fallback: Cls (global)")
  }

  @Test
  def cuCreationTest(): Unit = {
    val dummyFile = ProjectFile(testProjectPath, "geometry.h")

    // Class
    val circleClass = an.cuClass("shapes.Circle", dummyFile).get
    assertEquals("shapes", circleClass.packageName())
    assertEquals("Circle", circleClass.shortName())
    assertEquals("shapes.Circle", circleClass.fqName())

    val pointStruct = an.cuClass("Point", dummyFile).get
    assertEquals("", pointStruct.packageName())
    assertEquals("Point", pointStruct.shortName())
    assertEquals("Point", pointStruct.fqName())

    // Function/Method
    val getAreaMethod = an.cuFunction("shapes.Circle.getArea", dummyFile).get
    assertEquals("shapes", getAreaMethod.packageName())
    assertEquals("Circle.getArea", getAreaMethod.shortName())
    assertEquals("shapes.Circle.getArea", getAreaMethod.fqName())

    // Global function (assuming file-prefixing by c2cpg or parseFqName handling)
    val globalFunc = an.cuFunction("geometry_cpp.global_func", dummyFile).get
    assertEquals("geometry_cpp", globalFunc.packageName())
    assertEquals("global_func", globalFunc.shortName())
    assertEquals("geometry_cpp.global_func", globalFunc.fqName())

    // Namespaced function (not in a class)
    val anotherInShapes = an.cuFunction("shapes.another_in_shapes", dummyFile).get
    assertEquals("shapes", anotherInShapes.packageName())
    assertEquals("another_in_shapes", anotherInShapes.shortName())
    assertEquals("shapes.another_in_shapes", anotherInShapes.fqName())

    // Field
    val radiusField = an.cuField("shapes.Circle.radius", dummyFile).get
    assertEquals("shapes", radiusField.packageName())
    assertEquals("Circle.radius", radiusField.shortName())
    assertEquals("shapes.Circle.radius", radiusField.fqName())
  }

  @Test
  def isClassInProjectTest(): Unit = {
    assertTrue(an.isClassInProject("shapes.Circle"), "shapes.Circle should be in project")
    assertTrue(an.isClassInProject("Point"), "Point struct should be in project")
    // std types are typically external or filtered out by isClassInProject if it checks isExternal or file path
    assertFalse(an.isClassInProject("std::vector"), "std::vector should not be in project")
    assertFalse(an.isClassInProject("NonExistentClass"), "NonExistentClass should not be in project")
  }

  @Test
  def resolveMethodNameTest(): Unit = {
    assertEquals("shapes.Circle.getArea", an.resolveMethodName("shapes.Circle.getArea:double()"))
    assertEquals("global_func", an.resolveMethodName("global_func:void(int)"))
    // c2cpg uses <duplicate>N (N is an int), from io.joern.c2cpg.astcreation.Defines.DuplicateSuffix
    assertEquals("foo", an.resolveMethodName(s"foo${io.joern.c2cpg.astcreation.Defines.DuplicateSuffix}0:int()"))
  }

  @Test
  def sanitizeTypeTest(): Unit = {
    assertEquals("Circle", an.sanitizeType("shapes::Circle"))
    assertEquals("int", an.sanitizeType("int"))
    assertEquals("double", an.sanitizeType("double*"))
    assertEquals("MyType", an.sanitizeType("const MyNamespace::MyType&"))
    assertEquals("char", an.sanitizeType("char[]"))
    assertEquals("MyType", an.sanitizeType("MyNamespace.MyType"))
    assertEquals("MyType", an.sanitizeType("struct MyNamespace.MyType"))
    assertEquals("MyType", an.sanitizeType("MyType*"))
    assertEquals("MyType", an.sanitizeType("MyType[]"))
    assertEquals("MyType", an.sanitizeType("const MyType"))
    assertEquals("MyType", an.sanitizeType("MyNamespace.MyType*[]"))
    assertEquals("MyType", an.sanitizeType("enum MyNamespace::MyType"))
    assertEquals("MyType", an.sanitizeType("volatile MyType * const * restrict"))
    assertEquals(io.joern.x2cpg.Defines.Any, an.sanitizeType("const*")) // what should this be?
  }

  @Test
  def getSkeletonTest(): Unit = {
    val circleSkeletonOpt = an.getSkeleton("shapes.Circle")
    assertTrue(circleSkeletonOpt.isPresent, "Skeleton for shapes.Circle not found")
    val circleSkeleton = circleSkeletonOpt.get
    assertTrue(circleSkeleton.contains("class Circle {"), s"Skeleton was: $circleSkeleton") // Expect class based on methods/inheritance
    assertTrue(circleSkeleton.contains("double getArea() {...}"), s"Skeleton was: $circleSkeleton") // Match actual signature
    assertTrue(circleSkeleton.contains("double radius;"), s"Skeleton was: $circleSkeleton")

    val pointSkeletonOpt = an.getSkeleton("Point")
    assertTrue(pointSkeletonOpt.isPresent, "Skeleton for Point not found")
    val pointSkeleton = pointSkeletonOpt.get()
    // CPG might represent struct as class if it has methods or based on configuration
    assertTrue(pointSkeleton.contains("struct Point {") || pointSkeleton.contains("class Point {"), s"Skeleton was: $pointSkeleton")
    assertTrue(pointSkeleton.contains("void print() {...}"), s"Skeleton was: $pointSkeleton")
    assertTrue(pointSkeleton.contains("int x;"), s"Skeleton was: $pointSkeleton")
  }

  @Test
  def getMethodSourceTest(): Unit = {
    // Method in a namespace
    val getAreaOpt = an.getMethodSource("shapes.Circle.getArea")
    assertTrue(getAreaOpt.isPresent, "Could not find source for shapes.Circle.getArea")
    assertTrue(getAreaOpt.get.contains("return M_PI * radius * radius;"))

    // Global function. FQN depends on c2cpg (e.g., geometry_cpp.global_func)
    val globalFuncOpt = an.getMethodSource("geometry_cpp.global_func")
    assertTrue(globalFuncOpt.isPresent, "Could not find source for geometry_cpp.global_func")
    assertTrue(globalFuncOpt.get.contains("global_var = val;"))
  }


  @Test
  def getDeclarationsInFileTest(): Unit = {
    val fileH = Try(ProjectFile(testProjectPath, "geometry.h")).toOption
    assumeTrue(fileH.isDefined, "geometry.h project file could not be created")

    val declsInH = an.getDeclarationsInFile(fileH.get).asScala.toSet.map(_.fqName())

    assertTrue(declsInH.contains("shapes.Circle"), "Missing shapes.Circle class from geometry.h")
    assertTrue(declsInH.contains("Point"), "Missing Point struct from geometry.h")
    // Method declarations in headers might or might not be picked up as distinct CodeUnits by getDeclarationsInFile
    // depending on AbstractAnalyzer's logic and CPG structure.
    // For now, ensure main types are found.
    // assertTrue(declsInH.contains("shapes.Circle.getArea"), "Missing shapes.Circle.getArea method from geometry.h")
    assertTrue(declsInH.contains("Point.x"), "Missing Point.x field from geometry.h")


    val fileCpp = Try(ProjectFile(testProjectPath, "geometry.cpp")).toOption
    assumeTrue(fileCpp.isDefined, "geometry.cpp project file could not be created")
    val declsInCpp = an.getDeclarationsInFile(fileCpp.get).asScala.toSet.map(_.fqName())

    // Functions defined in geometry.cpp
    // FQNs for global functions can be tricky, e.g., "geometry_cpp.global_func"
    assertTrue(declsInCpp.contains("geometry_cpp.global_func"), "Missing global_func from geometry.cpp")
    assertTrue(declsInCpp.contains("geometry_cpp.uses_global_func"), "Missing uses_global_func from geometry.cpp")
    assertTrue(declsInCpp.contains("shapes.another_in_shapes"), "Missing shapes.another_in_shapes from geometry.cpp")
    // Ensure class/method definitions from cpp file are also found if they are primarily there
    assertTrue(declsInCpp.contains("shapes.Circle.getArea"), "Missing definition of shapes.Circle.getArea from geometry.cpp")
  }

  @Test
  def getUsesTest(): Unit = {
    // Test uses of global_func (e.g., geometry_cpp.global_func)
    val usesGlobalFunc = an.getUses("geometry_cpp.global_func").asScala.map(_.fqName()).toSet
    assertTrue(usesGlobalFunc.contains("main_cpp.main"), "main_cpp.main should use global_func")
    assertTrue(usesGlobalFunc.contains("geometry_cpp.uses_global_func"), "uses_global_func should use global_func")
    // main_calls_lib is in main.cpp, which is part of testcode-cpp, so it should be analyzed.
    assertTrue(usesGlobalFunc.contains("main_cpp.main_calls_lib"), "main_calls_lib should use global_func")

    // Test uses of shapes.Circle.getArea
    val usesGetArea = an.getUses("shapes.Circle.getArea").asScala.map(_.fqName()).toSet
    assertTrue(usesGetArea.contains("main_cpp.main"), "main_cpp.main should use shapes.Circle.getArea")
    assertTrue(usesGetArea.contains("shapes.another_in_shapes"), "another_in_shapes should use shapes.Circle.getArea")
    assertTrue(usesGetArea.contains("main_cpp.main_calls_lib"), "main_calls_lib should use shapes.Circle.getArea")

    // Test uses of class Point (as a type)
    val usesPoint = an.getUses("Point").asScala.map(_.fqName()).toSet
    assertTrue(usesPoint.contains("main_cpp.main"), "main_cpp.main function uses Point type")
  }

  @Test
  def searchDefinitionsTest(): Unit = {
    val circleDefs = an.searchDefinitions("Circle").asScala
    assertTrue(circleDefs.exists(cu => cu.fqName == "shapes.Circle" && cu.isClass), "Did not find shapes.Circle class")

    val globalDefs = an.searchDefinitions("global").asScala
    assertTrue(globalDefs.exists(cu => cu.fqName().endsWith("global_func") && cu.isFunction()), "Did not find global_func")

    val getAreaDefs = an.searchDefinitions("getArea").asScala
    assertTrue(getAreaDefs.exists(cu => cu.fqName == "shapes.Circle.getArea" && cu.isFunction()), "Did not find shapes.Circle.getArea")
  }

  @Test
  def getDefinitionTest(): Unit = {
    // Class
    val circleDefOpt = an.getDefinition("shapes.Circle")
    assertTrue(circleDefOpt.isPresent)
    assertEquals("shapes.Circle", circleDefOpt.get.fqName)
    assertTrue(circleDefOpt.get.isClass)

    // Method
    val getAreaDefOpt = an.getDefinition("shapes.Circle.getArea")
    assertTrue(getAreaDefOpt.isPresent)
    assertEquals("shapes.Circle.getArea", getAreaDefOpt.get.fqName)
    assertTrue(getAreaDefOpt.get.isFunction)

    // Global function (e.g., geometry_cpp.global_func)
    val globalFuncDefOpt = an.getDefinition("geometry_cpp.global_func")
    assertTrue(globalFuncDefOpt.isPresent, "Failed to find geometry_cpp.global_func")
    assertEquals("geometry_cpp.global_func", globalFuncDefOpt.get.fqName)
    assertTrue(globalFuncDefOpt.get.isFunction)

    // Field
    val radiusDefOpt = an.getDefinition("shapes.Circle.radius")
    assertTrue(radiusDefOpt.isPresent)
    assertEquals("shapes.Circle.radius", radiusDefOpt.get.fqName)
    assertFalse(radiusDefOpt.get.isClass)
    assertFalse(radiusDefOpt.get.isFunction) // It's a field, not a function

    // Non-existent
    val nonExistentOpt = an.getDefinition("NonExistent.Symbol")
    assertFalse(nonExistentOpt.isPresent)
  }

  @Test
  def getFunctionLocationTest(): Unit = {
    // Global function in geometry.cpp
    val globalFuncLoc = an.getFunctionLocation("geometry_cpp.global_func", java.util.List.of("val"))
    assertTrue(globalFuncLoc.code.contains("global_var = val;"))
    assertEquals(ProjectFile(testProjectPath, "geometry.cpp"), globalFuncLoc.file)

    // Method in class
    val getAreaLoc = an.getFunctionLocation("shapes.Circle.getArea", java.util.Collections.emptyList())
    assertTrue(getAreaLoc.code.contains("return M_PI * radius * radius;"))
    assertEquals(ProjectFile(testProjectPath, "geometry.cpp"), getAreaLoc.file) // Definition in .cpp
  }

  @Test
  def getMethodSourceFromPytorchTest(): Unit = {
    // Test retrieving source for a function in pytorch.cpp
    val funcFqn = "pytorch_cpp.start_index"
    val sourceOpt = an.getMethodSource(funcFqn)
    assertTrue(sourceOpt.isPresent, s"Could not find source for FQN: $funcFqn")

    val source = sourceOpt.get()
    // Check for the signature (first line of the function definition)
    val expectedSignaturePrefix = "inline int start_index(int out_idx, int out_len, int in_len) {"
    assertTrue(source.startsWith(expectedSignaturePrefix),
      s"Source for $funcFqn did not start with expected signature.\nExpected prefix: '$expectedSignaturePrefix'\nActual source:\n$source")

    // Check for a specific line in the body
    val expectedBodyContent = "std::floor((float)(out_idx * in_len) / out_len)"
    assertTrue(source.contains(expectedBodyContent),
      s"Source for $funcFqn did not contain expected content '$expectedBodyContent'.\nActual source:\n$source")
  }

  @Test
  def getSkeletonsPytorchTest(): Unit = {
    val file = ProjectFile(testProjectPath, "pytorch.cpp")
    val skeletons = an.getSkeletons(file)

    val expectedKey = CodeUnit.fn(file, "at.native", "start_index")
    val expectedSkeleton = "int start_index(int out_idx, int out_len, int in_len) {...}"

    assertEquals(1, skeletons.size(), s"Expected 1 skeleton, got ${skeletons.size()}. Skeletons: $skeletons")
    assertTrue(skeletons.containsKey(expectedKey), s"Skeletons map does not contain expected key '$expectedKey'. Actual keys: ${skeletons.keySet().asScala.mkString(", ")}")
    assertEquals(expectedSkeleton, skeletons.get(expectedKey), "Skeleton content mismatch.")
  }
}

File: app/src/test/java/io/github/jbellis/brokk/analyzer/CppAnalyzerTest.java
package io.github.jbellis.brokk.analyzer;

import static org.junit.jupiter.api.Assertions.*;

import io.github.jbellis.brokk.testutil.TestProject;
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;
import org.jetbrains.annotations.Nullable;
import org.junit.jupiter.api.AfterAll;
import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.api.Test;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

public class CppAnalyzerTest {

    private static final Logger logger = LoggerFactory.getLogger(CppAnalyzerTest.class);

    @Nullable
    private static CppAnalyzer analyzer;

    @Nullable
    private static TestProject testProject;

    @BeforeAll
    public static void setup() throws IOException {
        final var testPath =
                Path.of("src/test/resources/testcode-cpp").toAbsolutePath().normalize();
        assertTrue(Files.exists(testPath), "Test resource directory 'testcode-cpp' not found.");
        testProject = new TestProject(testPath, Languages.CPP_TREESITTER);
        logger.debug(
                "Setting up analyzer with test code from {}",
                testPath.toAbsolutePath().normalize());
        analyzer = new CppAnalyzer(testProject);
    }

    @AfterAll
    public static void teardown() {
        if (testProject != null) {
            testProject.close();
        }
    }

    @Test
    public void isEmptyTest() {
        assertFalse(analyzer.isEmpty());
    }

    private List<CodeUnit> getAllDeclarations() {
        return testProject.getAllFiles().stream()
                .flatMap(file -> analyzer.getDeclarationsInFile(file).stream())
                .collect(Collectors.toList());
    }

    @Test
    public void testNamespaceAnalysis() {
        var allDeclarations = getAllDeclarations();

        // Check for actual namespace declarations (MODULE type)
        var namespaceDeclarations = allDeclarations.stream()
                .filter(cu -> cu.kind() == CodeUnitType.MODULE)
                .collect(Collectors.toList());

        logger.debug("Found {} namespace declarations", namespaceDeclarations.size());
        namespaceDeclarations.forEach(cu -> logger.debug("Namespace: {} (type: {})", cu.shortName(), cu.kind()));

        // Should find specific namespace declarations
        assertTrue(namespaceDeclarations.size() >= 2, "Should find at least 2 namespace declarations");

        // Verify specific namespaces are detected
        var namespaceNames =
                namespaceDeclarations.stream().map(CodeUnit::shortName).collect(Collectors.toSet());

        assertTrue(namespaceNames.contains("graphics"), "Should find 'graphics' namespace");
        assertTrue(namespaceNames.contains("ui::widgets"), "Should find 'ui::widgets' nested namespace");
    }

    @Test
    public void testAnonymousNamespace() {
        var geometryFile = testProject.getAllFiles().stream()
                .filter(file -> file.absPath().toString().endsWith("geometry.cpp"))
                .findFirst()
                .orElseThrow(() -> new RuntimeException("geometry.cpp not found"));

        var allDeclarations = analyzer.getDeclarationsInFile(geometryFile);

        var allFunctions = allDeclarations.stream().filter(CodeUnit::isFunction).collect(Collectors.toList());

        var anonymousNamespaceFunctions = allDeclarations.stream()
                .filter(CodeUnit::isFunction)
                .filter(cu -> cu.shortName().contains("anonymous_helper")
                        || cu.shortName().contains("anonymous_void_func"))
                .collect(Collectors.toList());

        var functionNames = allFunctions.stream().map(cu -> cu.shortName()).collect(Collectors.toList());

        assertFalse(
                anonymousNamespaceFunctions.isEmpty(),
                "Should find function from anonymous namespace. Available functions: " + functionNames);

        var anonymousHelperFunction = anonymousNamespaceFunctions.stream()
                .filter(cu -> cu.identifier().equals("anonymous_helper"))
                .findFirst();

        assertTrue(
                anonymousHelperFunction.isPresent(), "Should find anonymous_helper function from anonymous namespace");

        var skeletons = analyzer.getSkeletons(geometryFile);
        var functionSkeletons = skeletons.entrySet().stream()
                .filter(entry -> entry.getKey().isFunction())
                .filter(entry -> entry.getKey().shortName().contains("anonymous_"))
                .collect(Collectors.toList());

        assertFalse(functionSkeletons.isEmpty(), "Should generate skeletons for anonymous namespace functions");
    }

    @Test
    public void testClassAnalysis() {
        var allDeclarations = getAllDeclarations();

        // Check for class declarations
        var classes = allDeclarations.stream()
                .filter(cu -> cu.kind() == CodeUnitType.CLASS)
                .filter(cu -> cu.shortName().contains("Circle")
                        || cu.shortName().contains("Renderer")
                        || cu.shortName().contains("Widget"))
                .collect(Collectors.toList());

        logger.debug("Found {} class declarations", classes.size());
        classes.forEach(cu -> logger.debug("Class: {} (type: {})", cu.shortName(), cu.kind()));

        assertTrue(classes.size() >= 2, "Should find at least Circle and other class declarations");
    }

    @Test
    public void testSkeletonOutput() {
        // Test that C++ skeletons include body placeholder for functions with bodies
        var geometryFile = testProject.getAllFiles().stream()
                .filter(file -> file.absPath().toString().endsWith("geometry.cpp"))
                .findFirst()
                .orElseThrow(() -> new RuntimeException("geometry.cpp not found"));

        var skeletons = analyzer.getSkeletons(geometryFile);

        logger.debug("=== All skeletons for geometry.cpp ===");
        skeletons.entrySet().forEach(entry -> {
            var codeUnit = entry.getKey();
            var skeleton = entry.getValue();
            logger.debug("CodeUnit: {} ({})", codeUnit.fqName(), codeUnit.kind());
            logger.debug("Skeleton: {}", skeleton);
            logger.debug("---");
        });

        // Look for functions that should have body placeholders
        var functionSkeletons = skeletons.entrySet().stream()
                .filter(entry -> entry.getKey().kind() == CodeUnitType.FUNCTION)
                .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue));

        assertFalse(functionSkeletons.isEmpty(), "Should find at least one function skeleton");

        // Check that function skeletons use {...} placeholder instead of actual body content
        for (var entry : functionSkeletons.entrySet()) {
            var skeleton = entry.getValue();
            var functionName = entry.getKey().fqName();

            logger.debug("Function: {} -> Skeleton: {}", functionName, skeleton);

            // The skeleton should contain {...} placeholder for functions with bodies
            if (functionName.contains("getArea")
                    || functionName.contains("print")
                    || functionName.contains("global_func")) {
                assertTrue(
                        skeleton.contains("{...}"),
                        "Function " + functionName + " should contain body placeholder {...}, but got: " + skeleton);
            }
        }
    }

    @Test
    public void testNestedClassSkeletonOutput() {
        var nestedFile = testProject.getAllFiles().stream()
                .filter(file -> file.absPath().toString().endsWith("nested.cpp"))
                .findFirst()
                .orElseThrow(() -> new RuntimeException("nested.cpp not found"));

        var skeletons = analyzer.getSkeletons(nestedFile);

        var outerClass = skeletons.entrySet().stream()
                .filter(entry -> entry.getKey().shortName().equals("Outer"))
                .findFirst();

        assertTrue(outerClass.isPresent(), "Should detect Outer class");

        var outerSkeleton = outerClass.get().getValue();
        assertTrue(
                outerSkeleton.contains("class Inner"), "Should detect nested Inner class within Outer class skeleton");

        var mainFunction = skeletons.keySet().stream()
                .filter(cu -> cu.fqName().contains("main"))
                .findFirst();

        assertTrue(mainFunction.isPresent(), "Should detect main function");
    }

    @Test
    public void testStructAnalysis() {
        var allDeclarations = getAllDeclarations();

        // Check for struct declarations
        var structs = allDeclarations.stream()
                .filter(cu -> cu.kind() == CodeUnitType.CLASS)
                .filter(cu -> cu.shortName().contains("Point"))
                .collect(Collectors.toList());

        logger.debug("Found {} struct declarations", structs.size());
        structs.forEach(cu -> logger.debug("Struct: {} (type: {})", cu.shortName(), cu.kind()));

        assertTrue(structs.size() >= 1, "Should find Point struct declaration");
    }

    @Test
    public void testGlobalDeclarations() {
        var allDeclarations = getAllDeclarations();

        var globalFunctions = allDeclarations.stream()
                .filter(cu -> cu.kind() == CodeUnitType.FUNCTION)
                .filter(cu -> cu.packageName().isEmpty())
                .filter(cu -> cu.fqName().contains("global"))
                .collect(Collectors.toList());

        assertTrue(
                globalFunctions.size() >= 2,
                "Should find at least 2 global functions: global_func and uses_global_func");
        assertTrue(
                globalFunctions.stream().anyMatch(cu -> cu.fqName().contains("global_func")),
                "Should find global_func declaration");
        assertTrue(
                globalFunctions.stream().anyMatch(cu -> cu.fqName().contains("uses_global_func")),
                "Should find uses_global_func declaration");

        var globalVariables = allDeclarations.stream()
                .filter(cu -> cu.kind() == CodeUnitType.FIELD)
                .filter(cu -> cu.packageName().isEmpty())
                .filter(cu -> cu.fqName().contains("global"))
                .collect(Collectors.toList());

        assertTrue(globalVariables.size() >= 1, "Should find at least 1 global variable: global_var");
        assertTrue(
                globalVariables.stream().anyMatch(cu -> cu.fqName().contains("global_var")),
                "Should find global_var declaration");
    }

    @Test
    public void testStructFieldsAndMethods() {
        var geometryFile = testProject.getAllFiles().stream()
                .filter(file -> file.absPath().toString().endsWith("geometry.h"))
                .findFirst()
                .orElseThrow(() -> new RuntimeException("geometry.h not found"));

        var skeletons = analyzer.getSkeletons(geometryFile);

        var pointStruct = skeletons.entrySet().stream()
                .filter(entry -> entry.getKey().shortName().equals("Point"))
                .findFirst();

        assertTrue(pointStruct.isPresent(), "Should find Point struct");

        var pointSkeleton = pointStruct.get().getValue();

        assertTrue(
                pointSkeleton.contains("int x") || pointSkeleton.contains("x"),
                "Point struct skeleton should contain field 'x'");
        assertTrue(
                pointSkeleton.contains("int y") || pointSkeleton.contains("y"),
                "Point struct skeleton should contain field 'y'");
    }

    @Test
    public void testEnumAnalysis() {
        var allDeclarations = getAllDeclarations();

        var enums = allDeclarations.stream()
                .filter(cu -> cu.kind() == CodeUnitType.CLASS)
                .filter(cu -> cu.shortName().contains("Color")
                        || cu.shortName().contains("BlendMode")
                        || cu.shortName().contains("Status")
                        || cu.shortName().contains("WidgetType"))
                .collect(Collectors.toList());

        assertTrue(enums.size() >= 1, "Should find enum declarations from advanced_features.h");
    }

    @Test
    public void testUnionAnalysis() {
        var allDeclarations = getAllDeclarations();

        var unions = allDeclarations.stream()
                .filter(cu -> cu.kind() == CodeUnitType.CLASS)
                .filter(cu -> cu.shortName().contains("Pixel") || cu.shortName().contains("DataValue"))
                .collect(Collectors.toList());

        assertTrue(unions.size() >= 1, "Should find union declarations from advanced_features.h");
    }

    @Test
    public void testNamespacePackageNaming() {
        var allDeclarations = getAllDeclarations();

        var classesWithNamespaces = allDeclarations.stream()
                .filter(cu -> cu.kind() == CodeUnitType.CLASS)
                .filter(cu -> !cu.packageName().isEmpty())
                .collect(Collectors.toList());

        var graphicsClasses = classesWithNamespaces.stream()
                .filter(cu -> cu.packageName().equals("graphics"))
                .collect(Collectors.toList());
        assertTrue(graphicsClasses.size() >= 2, "Should find classes in 'graphics' namespace");

        var nestedNamespaceClasses = classesWithNamespaces.stream()
                .filter(cu -> cu.packageName().equals("ui::widgets"))
                .collect(Collectors.toList());
        assertTrue(nestedNamespaceClasses.size() >= 1, "Should find classes in 'ui::widgets' nested namespace");

        var graphicsColorFound = classesWithNamespaces.stream()
                .anyMatch(cu ->
                        cu.packageName().equals("graphics") && cu.shortName().contains("Color"));
        assertTrue(graphicsColorFound, "Should find Color enum in graphics namespace");

        var graphicsRendererFound = classesWithNamespaces.stream()
                .anyMatch(cu ->
                        cu.packageName().equals("graphics") && cu.shortName().contains("Renderer"));
        assertTrue(graphicsRendererFound, "Should find Renderer class in graphics namespace");

        var widgetFound = classesWithNamespaces.stream()
                .anyMatch(cu ->
                        cu.packageName().equals("ui::widgets") && cu.shortName().contains("Widget"));
        assertTrue(widgetFound, "Should find Widget class in ui::widgets namespace");
    }

    @Test
    public void testTypeAliasAnalysis() {
        var allDeclarations = getAllDeclarations();

        var aliases = allDeclarations.stream()
                .filter(cu -> cu.kind() == CodeUnitType.CLASS)
                .filter(cu -> cu.shortName().contains("ColorValue")
                        || cu.shortName().contains("PixelBuffer")
                        || cu.shortName().contains("String")
                        || cu.shortName().contains("uint32_t"))
                .collect(Collectors.toList());

        logger.debug("Type aliases are not supported by current C++ TreeSitter grammar (expected: 0)");
    }

    @Test
    public void testFunctionAnalysis() {
        var allDeclarations = getAllDeclarations();

        var functions = allDeclarations.stream()
                .filter(cu -> cu.kind() == CodeUnitType.FUNCTION)
                .collect(Collectors.toList());

        assertTrue(functions.size() >= 3, "Should find multiple function declarations");
    }

    @Test
    public void testComprehensiveDeclarationCount() {
        var allDeclarations = getAllDeclarations();

        var byType = allDeclarations.stream().collect(Collectors.groupingBy(CodeUnit::kind, Collectors.counting()));

        assertTrue(allDeclarations.size() >= 10, "Should find at least 10 declarations with enhanced C++ constructs");

        assertTrue(byType.containsKey(CodeUnitType.CLASS), "Should find class-like declarations");
        assertTrue(byType.containsKey(CodeUnitType.FUNCTION), "Should find function declarations");
    }

    @Test
    public void testSpecificFileAnalysis() {
        var advancedFile = testProject.getAllFiles().stream()
                .filter(f -> f.toString().contains("advanced_features.h"))
                .findFirst();

        assertTrue(advancedFile.isPresent(), "Should find advanced_features.h test file");

        var declarations = analyzer.getDeclarationsInFile(advancedFile.get());

        assertTrue(
                declarations.size() >= 5,
                "Should find at least 5 declarations in advanced_features.h with enums, unions, classes, etc.");
    }

    @Test
    public void testAdvancedFeaturesSkeletonOutput() {
        var advancedFile = testProject.getAllFiles().stream()
                .filter(file -> file.absPath().toString().endsWith("advanced_features.h"))
                .findFirst()
                .orElseThrow(() -> new RuntimeException("advanced_features.h not found"));

        var skeletons = analyzer.getSkeletons(advancedFile);

        var namespaceSkeletons = skeletons.entrySet().stream()
                .filter(entry -> entry.getKey().kind() == CodeUnitType.MODULE)
                .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue));

        assertFalse(namespaceSkeletons.isEmpty(), "Should find at least one namespace");

        var graphicsNamespace = namespaceSkeletons.entrySet().stream()
                .filter(entry -> entry.getKey().fqName().equals("graphics"))
                .findFirst();

        assertTrue(graphicsNamespace.isPresent(), "Should find graphics namespace");
        String graphicsSkeleton = graphicsNamespace.get().getValue();

        assertTrue(
                graphicsSkeleton.contains("Color") || graphicsSkeleton.contains("Renderer"),
                "Graphics namespace should contain some declarations");
    }

    @Test
    public void testParseOnceCachingPerformance() {
        var geometryFile = testProject.getAllFiles().stream()
                .filter(file -> file.absPath().toString().endsWith("geometry.cpp"))
                .findFirst()
                .orElseThrow(() -> new RuntimeException("geometry.cpp not found"));

        logger.info("Initial cache stats: {}", analyzer.getCacheStatistics());

        // First call - should parse and cache the tree
        long startTime = System.nanoTime();
        var skeletons1 = analyzer.getSkeletons(geometryFile);
        long firstCallTime = System.nanoTime() - startTime;

        logger.info("After first getSkeletons() call:");
        logger.info("  - Time: {} ms", firstCallTime / 1_000_000.0);
        logger.info("  - Cache stats: {}", analyzer.getCacheStatistics());
        logger.info("  - Found {} skeletons", skeletons1.size());

        // Second call - should use cached tree (much faster)
        startTime = System.nanoTime();
        var skeletons2 = analyzer.getSkeletons(geometryFile);
        long secondCallTime = System.nanoTime() - startTime;

        logger.info("After second getSkeletons() call:");
        logger.info("  - Time: {} ms", secondCallTime / 1_000_000.0);
        logger.info("  - Cache stats: {}", analyzer.getCacheStatistics());
        logger.info("  - Found {} skeletons", skeletons2.size());

        // Verify results are identical
        assertEquals(skeletons1, skeletons2, "Results should be identical - caching works correctly");

        // Performance improvement validation
        assertTrue(skeletons1.size() > 0, "Should find at least one skeleton");

        if (secondCallTime > 0) {
            double improvement = (double) firstCallTime / secondCallTime;
            logger.info("Performance improvement ratio: {}x", String.format("%.2f", improvement));

            // Performance analysis (informational due to measurement variance at microsecond level)
            if (improvement > 1.2) {
                logger.info("✓ Significant performance improvement detected!");
            } else if (improvement >= 0.5) {
                logger.info("✓ Performance within expected range - caching is working");
            } else {
                logger.info("? Unexpected performance variance detected");
            }
        }
    }
}
